luarules/gadgets/unit_preaim.lua

```diff
@@ -14,19 +14,25 @@ if not gadgetHandler:IsSyncedCode() then
 	return
 end
 
+--use customparams.exclude_preaim = true, to exclude units from being able to pre-aim at targets almost within firing range.
+--this is a good idea for pop-up turrets so they don't prematurely reveal themselves.
+--also when proximityPriority is heavily biased toward far targets
+
 local weaponRange = {}
 local isPreaimUnit = {}
 for unitDefID, unitDef in pairs(UnitDefs) do
-	if not unitDef.canFly then
+	if not unitDef.canFly and not unitDef.customParams.exclude_preaim then
```

Comment:
Looks redundant on the unit, does nothing over applying it to all of a unit's weapons

Reply:
Sometimes you'll want it to be applied to specific weapons, other times to the entire unit. This method is flexible.

If you feel that's a "code smell", I can force you to *only* use the weapondef.customparam and leave no option for unitdef.customparam.

Some units will want this level of control.

Reply:
for example, the cortex flagship. Maybe it's problematic for the lasers to pre-aim due to proximity priority but the long range cannon is desirable to pre-aim. Or maybe I'm working with Evolving Commanders and the Armada EMP laser should be ready to blast for skirmishing purposes, but the backup close range lightening gun has weird proximityPriority interactions. Another example, maybe it turns out it's optimal for one weapon in a smart select arrangement from an array of possible weapons to be able to pre-aim while the other ones cannot. 

We just don't quite know yet and I'd rather have the flexibility for the balance bros to customparam one or all of the weapons of the unit with minimal code and effort.

Reply:
Put it purely on the weapons imo, that's what the tag applies to. Plus most units don't have multiple weapons.

Reply:
i would be careful about given flexibility to balance.
its best to leave something like this to just the gdt.

Reply:
I don't understand the importance of reduced flexibility. I'll do as you ask and make gadget less flexible and cool. Just know that Code Jesus is up there shaking his head disapprovingly from behind his coolguy sunglasses.